homeassistant/components/govee_ble/bluetooth_update_coordinator.py

```diff
@@ -0,0 +1,150 @@
+            dict[str, Any] | None,
+        ]
+        | None = None,
+    ) -> None:
+        """Initialize the dispatcher."""
+        self.hass = hass
+        self.logger = logger
+        self.name = name
+        self.parser_method = parser_method
+        self.address = address
+        self._listeners: dict[str | None, list[BluetoothListenerCallbackType]] = {}
+        self._cancel: CALLBACK_TYPE | None = None
+        self.last_update_success = True
+        self.last_exception: Exception | None = None
+
+    @callback
+    def async_setup(self) -> CALLBACK_TYPE:
+        """Start the callback."""
+        if self.parser_method is None:
+            raise NotImplementedError("Parser method not implemented")
+        return bluetooth.async_register_callback(
+            self.hass,
+            self._async_handle_bluetooth_event,
+            bluetooth.BluetoothCallbackMatcher(address=self.address),
+        )
+
+    @callback
+    def async_add_listener(
+        self, update_callback: BluetoothListenerCallbackType, key: str | None = None
+    ) -> Callable[[], None]:
+        """Listen for data updates."""
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self._listeners[key].remove(update_callback)
+            if not self._listeners[key]:
+                del self._listeners[key]
+
+        self._listeners.setdefault(key, []).append(update_callback)
+        return remove_listener
+
+    @callback
+    def async_update_listeners(self, data: dict[str, Any]) -> None:
+        """Update all registered listeners."""
+        # Dispatch to listeners without a filter key
+        if listeners := self._listeners.get(None):
+            for update_callback in listeners:
+                update_callback(data)
+
+        # Dispatch to listeners with a filter key
+        for key in data:
+            if listeners := self._listeners.get(key):
+                for update_callback in listeners:
+                    update_callback(data)
+
+    @callback
+    def _async_handle_bluetooth_event(
+        self,
+        service_info: bluetooth.BluetoothServiceInfo,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Handle a Bluetooth event."""
+        assert self.parser_method is not None
+        try:
+            parsed = self.parser_method(service_info, change)
+        except Exception as err:  # pylint: disable=broad-except
+            self.last_exception = err
+            self.last_update_success = False
+            self.logger.exception("Unexpected error update %s data: %s", self.name, err)
+        else:
+            if not self.last_update_success:
+                self.last_update_success = True
+                self.logger.info("Fetching %s data recovered", self.name)
+            if parsed:
+                self.async_update_listeners(parsed)
+
+
+class BluetoothCoordinatorEntity(entity.Entity):
```

Comment:
should it inherit from coordinator entity?

Reply:
I think we need a model where the data gets built up for that to work

Reply:
After additional refactoring they share is `available` and `should_poll`, and available likely needs to change